File: Sources/EwECustomPlugins/EwEValueChain/Data/Flow elements/Units/cUnit.vb
#Region " Imports "

Option Strict On
Imports System.ComponentModel
Imports System.Reflection
Imports EwEUtils.Utilities
Imports EwEUtils.Database
Imports EwECore
Imports ScientificInterfaceShared.Style

#End Region ' Imports

<TypeConverter(GetType(cPropertySorter)), _
    DefaultProperty("Name"), _
    Serializable()> _
Public MustInherit Class cUnit
    Inherits EwEUtils.Database.cEwEDatabase.cOOPStorable

    Protected Const sPROPCAT_GENERAL As String = "1. General"
    'Protected Const sPROPCAT_INTEGRATION As String = "2. Ecopath integration"
    Protected Const sPROPCAT_PRODUCTS As String = "2. Products ($/t)"
    Protected Const sPROPCAT_REVENUE As String = "3. Revenue ($/effort)"
    'Protected Const sPROPCAT_EFFORTSUBSIDIES As String = "4. Effort subsidies ($/effort)"
    Protected Const sPROPCAT_SUBSIDIES As String = "4. Subsidies ($/t)"
    Protected Const sPROPCAT_PAY As String = "5. Pay ($/t)"
    Protected Const sPROPCAT_SHARE As String = "5. Share (% revenue)"
    'Protected Const sPROPCAT_EFFORTCOST As String = "6. Effort cost ($/effort)"
    Protected Const sPROPCAT_INPUTCOST As String = "6. Input cost ($/t)"
    Protected Const sPROPCAT_TAXES As String = "7. Taxes ($/t)"
    Protected Const sPROPCAT_SOCIAL As String = "8. Social (#/t)"


    ''' <summary>Index of the unit, which this unit needs to store its values in the Results object</summary>
    Private m_iSequence As Integer = 0
    ''' <summary>List of input variables that this unit needs in order to perform its calculations.</summary>
    Private m_lReceivedInputs As New List(Of cInput)
    ''' <summary>Name of the unit</summary>
    Private m_strName As String
    ''' <summary>Nationality of a unit.</summary>
    Private m_iNationality As Integer
    ''' <summary>Zhe ceur</summary>
    Private m_core As cCore = Nothing

#Region " Constructor "

    ''' -----------------------------------------------------------------------
    ''' <summary>
    ''' 
    ''' </summary>
    ''' -----------------------------------------------------------------------
    Public Sub New()
        MyBase.New()
    End Sub

#End Region ' Constructor

#Region " Links "

    ''' <summary>Units that receive outputs from this unit.</summary>
    Private m_llinkOutput As New List(Of cLink)
    ''' <summary>Units that provide inputs for this unit.</summary>
    Private m_llinkInput As New List(Of cLink)

    Public Function LinkOutCount() As Integer
        Return Me.m_llinkOutput.Count
    End Function

    Public Function LinkOut(ByVal iIndex As Integer) As cLink
        Return Me.m_llinkOutput(iIndex)
    End Function

    Public Sub AddLink(ByVal link As cLink)

        ' Sanity check
        Debug.Assert(Object.ReferenceEquals(link.Source, Me))

        Me.m_llinkOutput.Add(link)
        link.Target.AddInputLink(link)
    End Sub

    Public Sub RemoveLink(ByVal link As cLink)
        Me.m_llinkOutput.Remove(link)
        link.Target.RemoveInputLink(link)
    End Sub

    Public Function LinkInCount() As Integer
        Return Me.m_llinkInput.Count
    End Function

    Public Function LinkIn(ByVal iIndex As Integer) As cLink
        Return Me.m_llinkInput(iIndex)
    End Function

    Protected Sub AddInputLink(ByVal link As cLink)

        ' Sanity check
        Debug.Assert(Object.ReferenceEquals(link.Target, Me))

        Me.m_llinkInput.Add(link)
    End Sub

    Protected Sub RemoveInputLink(ByVal link As cLink)
        Me.m_llinkInput.Remove(link)
    End Sub

    Public Function IsLoop(ByVal unit As cUnit) As Boolean

        ' Linked to self?
        Dim bIsLoop As Boolean = Object.ReferenceEquals(unit, Me)

        'Console.WriteLine("{0}:{1}={2}", Me.Name, unit.Name, bIsLoop)

        ' If no loop yet
        If Not bIsLoop Then
            ' Follow each output link
            For Each link As cLink In Me.m_llinkOutput
                ' See the target link is the requesting unit
                If link.Target.IsLoop(unit) Then bIsLoop = True : Exit For
            Next link
        End If

        Return bIsLoop
    End Function

    Public Function HasTarget(ByVal unit As cUnit) As Boolean

        ' Follow each output link
        For Each link As cLink In Me.m_llinkOutput
            ' See the target link is the requesting unit
            If Object.ReferenceEquals(link.Target, unit) Then Return True
        Next link
        Return False

    End Function

#End Region ' Links 

#Region " Running "

    ''' -----------------------------------------------------------------------
    ''' <summary>
    ''' Initialize the unit for a new Ecosim or Ecospace run.
    ''' </summary>
    ''' <param name="core">The EwE core that this run is performed onto.</param>
    ''' <param name="iSequence">The sequence number to assign to this unit for the run.</param>
    ''' -----------------------------------------------------------------------
    Public Overridable Sub InitRun(ByVal core As cCore, ByVal iSequence As Integer)
        Me.Sequence = iSequence
        Me.Core = core
    End Sub

    ''' -----------------------------------------------------------------------
    ''' <summary>
    ''' Initialize the unit for running a chain.
    ''' </summary>
    ''' -----------------------------------------------------------------------
    Public Overridable Sub Clear()
        ' Clear all pending inputs
        Me.m_lReceivedInputs.Clear()
    End Sub

    ''' -----------------------------------------------------------------------
    ''' <summary>
    ''' Calculate the economics for this unit.
    ''' </summary>
    ''' <param name="results"></param>
    ''' <param name="input"></param>
    ''' <param name="iTimeStep"></param>
    ''' -----------------------------------------------------------------------
    Public Overridable Sub Process(ByVal results As cResults, _
                                   ByVal input As cInput, _
                                   ByVal iTimeStep As Integer, _
                                   ByVal iFleet As Integer)

        Dim inputTotal As cInput = Nothing
        Dim sTotalOutputBiomass As Single = 0
        Dim sTotalOutputValue As Single = 0
        Dim sValuePerTon As Single = 0.0!

        ' Store received values
        Me.m_lReceivedInputs.Add(input)

        ' At least expected inputs received?
        If (Me.m_lReceivedInputs.Count >= Me.LinkInCount()) Then

            ' #Yes: Process!

            ' Combine all inputs
            inputTotal = Me.SumInputs(Me.m_lReceivedInputs)

            ' Determine outgoing biomass
            For Each link As cLink In Me.m_llinkOutput
                ' Determing output biomass for a single link
                Dim sOutputBiomass As Single = link.BiomassRatio * inputTotal.Tons

                sValuePerTon = link.ValuePerTon
                ' Is default link value?
                If (sValuePerTon = 1.0!) And (inputTotal.Tons <> 0.0!) Then
                    ' #Yes: use aggregated input value
                    sValuePerTon = inputTotal.Value / inputTotal.Tons
                End If

                sTotalOutputValue += sValuePerTon * sOutputBiomass
                sTotalOutputBiomass += sOutputBiomass
            Next

            results.StoreFleetContribution(iFleet, Me, iTimeStep, inputTotal.Value)

            ' No fleet specified?
            If iFleet = 0 Then
                ' #Yes: make all calculations
                Me.Calculate(results, _
                    inputTotal.Tons, inputTotal.Value, _
                    sTotalOutputBiomass, sTotalOutputValue, iTimeStep)
            End If


            ' Pass biomass to all targets in the flow
            For Each outputLink As cLink In Me.m_llinkOutput
                ' Pass resulting data to the next unit in the flow, and tell it to process

                ' Get system-defined value per ton
                sValuePerTon = outputLink.ValuePerTon
                ' Is default value?
                If (sValuePerTon = 1.0!) And (inputTotal.Tons <> 0.0!) Then
                    ' #Yes: use aggregated input value
                    sValuePerTon = inputTotal.Value / inputTotal.Tons
                End If

                outputLink.Target.Process(results, _
                        New cInput(inputTotal.Tons * outputLink.BiomassRatio, _
                                   inputTotal.Tons * outputLink.BiomassRatio * sValuePerTon), iTimeStep, iFleet)
            Next outputLink
        End If

    End Sub

    Protected Overridable Function SumInputs(ByVal lInputs As List(Of cInput)) As cInput
        Dim sTonsTotal As Single = 0.0
        Dim sValueTotal As Single = 0.0
        For Each input As cInput In lInputs
            sTonsTotal += input.Tons
            If (input.CustomValuePerTon <> 1) Then
                sValueTotal += (input.Tons * input.CustomValuePerTon)
            Else
                sValueTotal += input.Value
            End If
        Next
        Return New cInput(sTonsTotal, sValueTotal)
    End Function

    ''' <summary>
    ''' Make all calculations.
    ''' </summary>
    ''' <param name="results">The results object to store calculation results in.</param>
    Protected Overridable Function Calculate(ByVal results As cResults, _
            ByVal sInputBiomass As Single, ByVal sInputValue As Single, _
            ByVal sOutputBiomass As Single, ByVal sOutputValue As Single, _
            ByVal iTimeStep As Integer) As Boolean

        ' All good
        Return True

    End Function

#End Region ' Run

#Region " Copy / paste "

    Public Overrides Sub CopyFrom(ByVal obj As cEwEDatabase.cOOPStorable)
        Me.AllowEvents = False
        MyBase.CopyFrom(obj)
        Me.AllowEvents = True
    End Sub

#End Region ' Copy / paste

#Region " Properties "

    <Browsable(False)> _
    Public Property Sequence() As Integer
        Get
            Return Me.m_iSequence
        End Get
        Private Set(ByVal value As Integer)
            Me.m_iSequence = value
        End Set
    End Property

    <Browsable(False)> _
    Protected Property Core() As cCore
        Get
            Return Me.m_core
        End Get
        Private Set(ByVal value As cCore)
            Me.m_core = value
        End Set
    End Property

    <Browsable(False)> _
    Public MustOverride ReadOnly Property UnitType() As cUnitFactory.eUnitType

    <Browsable(False)> _
    Public Overridable ReadOnly Property HasError() As Boolean
        Get
            Return False
        End Get
    End Property

    <Browsable(True), _
        Category(sPROPCAT_GENERAL), _
        DisplayName("Name"), _
        Description("Name of this unit"), _
        cPropertySorter.PropertyOrder(1)> _
    Public Overridable Property Name() As String
        Get
            Return m_strName
        End Get
        Set(ByVal value As String)
            Me.m_strName = value
            Me.SetChanged()
        End Set
    End Property

    <Browsable(True), _
        Category(sPROPCAT_GENERAL), _
        DisplayName("Nationality"), _
        Description("Nationality of this unit"), _
        cPropertySorter.PropertyOrder(3)> _
    Public Overridable Property Nationality() As Integer
        Get
            Return Me.m_iNationality
        End Get
        Set(ByVal value As Integer)
            Me.m_iNationality = value
            Me.SetChanged()
        End Set
    End Property

    <Browsable(True), _
        Category(sPROPCAT_GENERAL), _
        DisplayName("Category"), _
        Description("Category to which this unit belongs"), _
        cPropertySorter.PropertyOrder(2)> _
    Public MustOverride ReadOnly Property Category() As String

    <Browsable(False)> _
    Public Overridable ReadOnly Property Style() As cStyleGuide.eStyleFlags
        Get
            Return cStyleGuide.eStyleFlags.OK
        End Get
    End Property

#End Region ' Properties

End Class

File: Sources Internal/ValueChain/ValueChain/Data/Flow elements/Units/cUnit.cs
// ===============================================================================
// This file is part of Ecopath with Ecosim (EwE)
//
// EwE is free software: you can redistribute it and/or modify it under the terms
// of the GNU General Public License version 2 as published by the Free Software 
// Foundation.
//
// EwE is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; 
// without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR 
// PURPOSE. See the GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along with EwE.
// If not, see <http://www.gnu.org/licenses/gpl-2.0.html>. 
//
//
// Copyright 1991- 
//    Ecopath International Initiative, Barcelona, Spain
// ===============================================================================

#region  Imports 

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Diagnostics;

namespace ValueChain
{

    #endregion

    [TypeConverter(typeof(cPropertySorter))]
    [DefaultProperty("Name")]
    [Serializable()]
    public abstract class cUnit : cValueChainEntity
    {

        protected const string sPROPCAT_GENERAL = "01. General";
        protected const string sPROPCAT_VALIDATION = "02. Validation";
        protected const string sPROPCAT_PRODUCTS = "03. Products ($/t)";
        protected const string sPROPCAT_REVENUE = "04. Revenue ($/effort)";
        protected const string sPROPCAT_SUBSIDIES = "05. Subsidies ($/t)";
        protected const string sPROPCAT_PAY = "06. Pay ($/t)";
        protected const string sPROPCAT_SHARE = "07. Share (% revenue)";
        protected const string sPROPCAT_INPUTCOST = "08. Input cost ($/t)";
        protected const string sPROPCAT_TAXES = "09. Taxes ($/t)";
        protected const string sPROPCAT_SOCIAL = "10. Social (#/t)";

        /// <summary>Index of the unit, which this unit needs to store its values in the Results object</summary>
        private int m_iSequence = 0;
        /// <summary>List of input variables that this unit needs in order to perform its calculations.</summary>
        protected List<cInput> m_lReceivedInputs = new List<cInput>();
        /// <summary>Name of the unit</summary>
        private string m_strName;
        /// <summary>Local name of the unit.</summary>
        private string m_strNameLocal;
        /// <summary>Nationality of a unit.</summary>
        private int m_iNationality;

        private bool m_bCanCompute = false;
        private bool m_bRunStarted = false;

        /// <summary>Units that receive outputs from this unit.</summary>
        protected List<cLink> m_llinkOutput = new List<cLink>();
        /// <summary>Units that provide inputs for this unit.</summary>
        protected List<cLink> m_llinkInput = new List<cLink>();

        #region  Constructor 

        /// -----------------------------------------------------------------------
    /// <summary>
    /// 
    /// </summary>
    /// -----------------------------------------------------------------------
        public cUnit() : base()
        {
        }

        #endregion

        #region  Links 

        public int LinkOutCount()
        {
            return m_llinkOutput.Count;
        }

        public cLink LinkOut(int iIndex)
        {
            return m_llinkOutput[iIndex];
        }

        /// <summary>
    /// Get all links directly linking to a target.
    /// </summary>
    /// <param name="unitTarget"></param>
    /// <returns></returns>
        public cLink[] Links(cUnit unitTarget)
        {
            var lLinks = new List<cLink>();
            foreach (cLink link in m_llinkOutput)
            {
                if (ReferenceEquals(link.Target, unitTarget))
                {
                    lLinks.Add(link);
                }
            }
            return lLinks.ToArray();
        }

        public void AddLink(cLink link)
        {

            // Sanity check
            Debug.Assert(ReferenceEquals(link.Source, this));

            m_llinkOutput.Add(link);
            link.Target.AddInputLink(link);
        }

        public void RemoveLink(cLink link)
        {
            m_llinkOutput.Remove(link);
            link.Target.RemoveInputLink(link);
        }

        public int LinkInCount()
        {
            return m_llinkInput.Count;
        }

        public cLink LinkIn(int iIndex)
        {
            return m_llinkInput[iIndex];
        }

        protected void AddInputLink(cLink link)
        {
            // Sanity check
            Debug.Assert(ReferenceEquals(link.Target, this));
            m_llinkInput.Add(link);
            UpdateComputeStatus();
        }

        protected void RemoveInputLink(cLink link)
        {
            m_llinkInput.Remove(link);
            UpdateComputeStatus();
        }

        public bool IsLoop(cUnit unit)
        {

            // Linked to self?
            bool bIsLoop = ReferenceEquals(unit, this);

            // If no loop yet
            if (!bIsLoop)
            {
                // Follow each output link
                foreach (cLink link in m_llinkOutput)
                {
                    // See the target link is the requesting unit
                    if (link.Target.IsLoop(unit))
                    {
                        bIsLoop = true;
                        break;
                    }
                }
            }

            return bIsLoop;
        }

        // Public Overridable Function HasTarget(unit As cUnit) As Boolean

        // ' Follow each output link
        // For Each link As cLink In Me.m_llinkOutput
        // ' See the target link is the requesting unit
        // If Object.ReferenceEquals(link.Target, unit) Then Return True
        // Next link
        // Return False

        // End Function

        #endregion

        #region  Running 

        /// -----------------------------------------------------------------------
    /// <summary>
    /// Initialize the unit for a new Ecosim or Ecospace run.
    /// </summary>
    /// <param name="iSequence">The sequence number to assign to this unit for the run.</param>
    /// -----------------------------------------------------------------------
        internal virtual void InitRun(int iSequence)
        {
            Sequence = iSequence;
            m_bRunStarted = true;
            Clear();
        }

        /// -----------------------------------------------------------------------
    /// <summary>
    /// Initialize the unit for running a chain.
    /// </summary>
    /// -----------------------------------------------------------------------
        internal virtual void Clear()
        {
            // Clear all pending inputs
            m_lReceivedInputs.Clear();
        }

        /// -----------------------------------------------------------------------
    /// <summary>
    /// Calculate the economics for this unit.
    /// </summary>
    /// <param name="results"></param>
    /// <param name="input"></param>
    /// <param name="iTimeStep"></param>
    /// <param name="iUnit">The unit to aggregate by.</param>
    /// -----------------------------------------------------------------------
        public virtual void Process(cValueChainResults results, cInput input, int iTimeStep, int iUnit)
        {

            float sTotalOutputBiomass = 0f;
            float sTotalOutputValue = 0f;
            float sValuePerTon = 0.0f;

            // Store received values
            m_lReceivedInputs.Add(input);

            // Sanity check
            Debug.Assert(m_lReceivedInputs.Count <= LinkInCount());

            // At least expected inputs received?
            if (m_lReceivedInputs.Count == LinkInCount())
            {

                // #Yes: Process combined inputs
                input = ProcessAndSumInputs(m_lReceivedInputs, results);

                // Store the amount that each fleet contributes to the total
                results.StoreContribution(iUnit, this, iTimeStep, input.Value, input.Tons);

                // Determine outgoing biomass
                foreach (cLink link in m_llinkOutput)
                {
                    // Determine output biomass for a single link
                    float sOutputBiomass = link.BiomassRatio * input.Tons;
                    float sOutputValue = 0f;

                    if (link.ValuePerTon != 1.0f & link.ValuePerTon != 0f | input.Tons == 0f)
                    {
                        sOutputValue = link.ValuePerTon * sOutputBiomass;
                    }
                    else
                    {
                        sOutputValue = input.Value / input.Tons * link.ValueRatio * sOutputBiomass;
                    }

                    sTotalOutputBiomass += sOutputBiomass;
                    sTotalOutputValue += sOutputValue;

                    link.Target.Process(results, new cInput(this, sOutputBiomass, sOutputValue), iTimeStep, iUnit);

                }

                // Running for all fleet?
                if (iUnit == 0)
                {
                    // #Yes: make all calculations. Calculations are not necessary when running for individual fleets
                    // where only transfer ratios are collected.
                    Calculate(results, input.Tons, input.Value, sTotalOutputBiomass, sTotalOutputValue, iTimeStep);
                }

            }

        }

        protected cInput ProcessAndSumInputs(List<cInput> lInputs, cValueChainResults results)
        {

            float sTonsTotal = 0.0f;
            float sValueTotal = 0.0f;

            foreach (cInput input in lInputs)
            {
                if (input.Tons > 0f)
                {

                    sTonsTotal += input.Tons;
                    sValueTotal += input.Value;

                    results.get_FlowsBiomass(input.Source.Sequence, Sequence);

                }
            }
            return new cInput(null, sTonsTotal, sValueTotal);

        }

        /// <summary>
    /// Make all calculations.
    /// </summary>
    /// <param name="results">The results object to store calculation results in.</param>
        protected virtual bool Calculate(cValueChainResults results, float sInputBiomass, float sInputValue, float sOutputBiomass, float sOutputValue, int iTimeStep)
        {

            // All good
            return true;

        }

        /// <summary>
    /// Assess whether a unit is ready to compute, e.g. when all its
    /// inputs are (in)directly connected to EwE model data.
    /// </summary>
        private void UpdateComputeStatus()
        {

            // Check if all input links can compute
            bool bCanCompute = true;
            bool bHasInputs = false;
            foreach (cLink LinkIn in m_llinkInput)
            {
                bCanCompute = bCanCompute & LinkIn.Source.CanCompute;
                bHasInputs = true;
            }
            bCanCompute = bCanCompute & bHasInputs;

            // No changes? Abort
            if (bCanCompute == CanCompute)
                return;

            m_bCanCompute = bCanCompute;

            foreach (cLink linkOut in m_llinkOutput)
                linkOut.Target.UpdateComputeStatus();

        }

        #endregion

        #region  Properties 

        public override string ToString()
        {
            return Name;
        }

        [Browsable(false)]
        public int Sequence
        {
            get
            {
                return m_iSequence;
            }
            private set
            {
                m_iSequence = value;
            }
        }

        [Browsable(false)]
        public abstract cUnitFactory.eUnitType UnitType { get; }

        [Browsable(false)]
        public virtual bool HasError
        {
            get
            {
                return false;
            }
        }

        // <Browsable(False)> _
        // Public Overridable ReadOnly Property Style() As cStyleGuide.eStyleFlags
        // Get
        // Return cStyleGuide.eStyleFlags.OK
        // End Get
        // End Property

        [Browsable(false)]
        public virtual bool CanCompute
        {
            get
            {
                return m_bCanCompute;
            }
        }

        [Browsable(false)]
        public virtual bool IsRunError
        {
            get
            {
                // Return if all results received OR when not ready to run yet
                return m_lReceivedInputs.Count < m_llinkInput.Count & m_bRunStarted;
            }
        }

        #region  General 

        [Browsable(true)]
        [Category(sPROPCAT_GENERAL)]
        [DisplayName("Name")]
        [Description("Name of this unit")]
        [cPropertySorter.PropertyOrder(1)]
        public virtual string Name
        {
            get
            {
                return m_strName;
            }
            set
            {
                m_strName = value;
                SetChanged();
            }
        }

        [Browsable(true)]
        [Category(sPROPCAT_GENERAL)]
        [DisplayName("Category")]
        [Description("Category to which this unit belongs")]
        [cPropertySorter.PropertyOrder(2)]
        public abstract string Category { get; }

        [Browsable(true)]
        [Category(sPROPCAT_VALIDATION)]
        [DisplayName("Biomass ratio")]
        [Description("Total biomass ratio passed out of this unit")]
        [cPropertySorter.PropertyOrder(8)]
        public virtual string BiomassRatio
        {
            get
            {
                float sTot = 0f;
                for (int i = 0, loopTo = LinkOutCount() - 1; i <= loopTo; i++)
                    sTot += LinkOut(i).BiomassRatio;
                return sTot.ToString();
            }
        }

        [Browsable(true)]
        [Category(sPROPCAT_GENERAL)]
        [DisplayName("Nationality")]
        [Description("Nationality of this unit")]
        [cPropertySorter.PropertyOrder(4)]
        public virtual int Nationality
        {
            get
            {
                return m_iNationality;
            }
            set
            {
                m_iNationality = value;
                SetChanged();
            }
        }

        [Browsable(true)]
        [Category(sPROPCAT_GENERAL)]
        [DisplayName("Name (local)")]
        [Description("Local name of this unit")]
        [cPropertySorter.PropertyOrder(5)]
        public virtual string NameLocal
        {
            get
            {
                return m_strNameLocal;
            }
            set
            {
                m_strNameLocal = value;
                SetChanged();
            }
        }

        [Browsable(false)]
        public abstract bool IsDefault { get; }

        #endregion

        #endregion

    }
}